conda/gateways/connection/session.py

```diff
@@ -60,6 +64,61 @@ def close(self):
         raise NotImplementedError()
 
 
+def get_channel_name_from_url(url: str) -> str | None:
+    """
+    Given a URL, determine the channel it belongs to and return its name.
+    """
+    for channel in get_channel_objs(context):
+        for base_url in channel.base_urls:
+            if url.lower().startswith(base_url):
+                return channel.canonical_name
```

Comment:
hmm, I would have thought that those would appear in channels too.. 😦

If that's not the case, then I'm not sure how we would support that...

Reply:
@kenodegard,

I thought about this some more today, and I don't believe this is a show stopper. I did some digging around and this is what I found...

When you pass something to `MatchSpec` that looks like this: `http://localhost/channel::package`, the URL is stripped away from the package name, which indicates to me that it is not meant to be used with URLs.

I am pretty sure this is only supposed to be used with channel names, and follow this approach I was able to get everything working. Below is how I set up my `.condarc` file by making use of the `channel_alias` setting:

```yaml
channels:
  - channel

channel_alias: http://localhost:8041

channel_settings:
  - channel: channel
    auth: conda-auth-oauth2
    login_url: http://localhost:8041/login
```

When you run the following command, everything works as expected (i.e. the correct auth handler is used):

```
conda create -n tester channel::travis_test
```

I actually think that this use case closely mirrors what actual users with private channels will be doing.

Thanks for bringing this up though. This was an important use case to think through. Please let me know if I have missed anything else important to consider.

Reply:
`https://conda.anaconda.org/conda-forge/noarch::nose2` is a valid package spec:

`conda create -n scratch https://conda.anaconda.org/conda-forge/noarch::nose2`

But even if thats found to be irrelevant I still think doing `conda.models.channel.Channel.from_url(url).canonical_name` will be more robust instead of using this custom `get_channel_name_from_url(url)` implementation